VMEncryption/main/handle.py

```diff
@@ -270,6 +271,89 @@ def get_protected_settings():
+        hutil.save_seq()
+        message = "Failed to update encryption settings with error: {0}, stack trace: {1}".format(e, traceback.format_exc())
+        hutil.do_exit(exit_code=CommonVariables.missing_dependency,
+                      operation='UpdateEncryptionSettingsLuks2Header',
+                      status=CommonVariables.extension_error_status,
+                      code=str(CommonVariables.missing_dependency),
+                      message=message)
+    try:
+        public_setting = get_public_settings()
+        encryption_config = EncryptionConfig(encryption_environment, logger)
+        extension_parameter = ExtensionParameter(hutil, logger, DistroPatcher, encryption_environment, get_protected_settings(), public_setting)
+        disk_util = DiskUtil(hutil=hutil, patching=DistroPatcher, logger=logger, encryption_environment=encryption_environment)
+        bek_util = BekUtil(disk_util, logger,encryption_environment)
+        device_items = disk_util.get_device_items(None)
+        for device_item in device_items:
+            device_item_path = disk_util.get_device_path(device_item.name)
+            if not disk_util.is_luks_device(device_item_path,None):
+                logger.log("Not a LUKS device, device path: {0}".format(device_item_path))
+                continue
+            #restoring the token data to type Azure_Disk_Encryption
+            disk_util.restore_luks2_token(device_name=device_item.name)
+            logger.log("Reading passphrase from LUKS2 header, device name: {0}".format(device_item.name))
+            #keep token copy for manual recovery
+            ade_token_id = disk_util.get_token_id(header_or_dev_path=device_item_path,token_name=CommonVariables.AzureDiskEncryptionToken)
+            if not ade_token_id:
+                logger.log("token type: Azure_Disk_Encryption not found for device {0}".format(device_item.name))
+                continue
+            data = disk_util.read_token(device_name=device_item.name,token_id=ade_token_id)
+            #writing to backup token
+            data['type']=CommonVariables.AzureDiskEncryptionBackUpToken
+            disk_util.import_token_data(device_path=device_item_path,token_data=data,token_id=CommonVariables.cvm_ade_vm_encryption_backup_token_id)
+            #get the unwrapped passphrase from LUKS2 header.
+            passphrase=disk_util.export_token(device_name=device_item.name)
+            #remove token
+            disk_util.remove_token(device_name=device_item.name,token_id=ade_token_id)
+            if not passphrase:
+                logger.log("No passphrase in LUKS2 header, device name: {0}".format(device_item.name))
+                continue
+            logger.log("Updating wrapped passphrase to LUKS2 header with current public setting. device name {0}".format(device_item.name))
+            #protect passphrase before updating to LUKS2 is done in import_token
+            temp_keyfile = tempfile.NamedTemporaryFile(delete=False)
```

Comment:
I see many occurrences of updating LUKS2 using temp file. Can't this be refactored as a separate function ?

Reply:
I will think on it more and do it later.